src/c2goto/library/pthread_lib.c

```diff
@@ -483,23 +483,39 @@ __ESBMC_HIDE:;
   return 0;
 }
 
+__attribute__((annotate("__ESBMC_inf_size"))) 
+const void *__ESBMC_thread_keys[];
+
+__attribute__((annotate("__ESBMC_inf_size"))) 
+void (*__ESBMC_thread_key_dtors[])(void *);
+
+unsigned long __ESBMC_next_thread_key;
+
 int pthread_key_create(pthread_key_t *key, void (*destructor)(void *))
 {
 __ESBMC_HIDE:;
-  // TODO
+  __ESBMC_thread_keys[__ESBMC_next_thread_key] = 0;
+  __ESBMC_thread_key_dtors[__ESBMC_next_thread_key] = destructor;
+  *key = __ESBMC_next_thread_key++;
+  return 0;
+}
+
+int pthread_key_delete(pthread_key_t key)
```

Comment:
_"Any attempt to use key following the call to pthread_key_delete() results in undefined behavior."_ (see [pthread_key_delete documentation](https://linux.die.net/man/3/pthread_key_delete)). We should make explicitly in the header of this function that we do not consider that behavior.

We could consider that all non-initialized positions in `__ESBMC_thread_keys` (during creation) are an specific (-1) object. Something like that in `pthread_key_create`,
```c
size_t nondet_pos;
__ESBMC_assume(nondet_pos != key);
__ESBMC_thread_keys[nondet_pos] = -1;
```

If this works, we could improve this model:
```c
if(__ESBMC_thread_keys[key] == -1) {
    return EINVAL;
}
__ESBMC_thread_keys[key] = 0;
return 0;
```